File: stripe/src/main/java/com/stripe/android/model/ConfirmPaymentIntentParams.java
package com.stripe.android.model;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.stripe.android.ObjectBuilder;
import com.stripe.android.utils.ObjectUtils;

import java.util.AbstractMap;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

@SuppressWarnings("WeakerAccess")
public final class ConfirmPaymentIntentParams implements ConfirmStripeIntentParams {

    public static final String API_PARAM_SOURCE_DATA = "source_data";
    public static final String API_PARAM_PAYMENT_METHOD_DATA = "payment_method_data";

    static final String API_PARAM_SOURCE_ID = "source";
    static final String API_PARAM_SAVE_PAYMENT_METHOD = "save_payment_method";

    @Nullable private final PaymentMethodCreateParams mPaymentMethodCreateParams;
    @Nullable private final String mPaymentMethodId;
    @Nullable private final SourceParams mSourceParams;
    @Nullable private final String mSourceId;

    @Nullable private final Map<String, Object> mExtraParams;
    @NonNull private final String mClientSecret;
    @Nullable private final String mReturnUrl;

    private final boolean mSavePaymentMethod;
    private final boolean mUseStripeSdk;

    private ConfirmPaymentIntentParams(@NonNull Builder builder) {
        mClientSecret = builder.mClientSecret;
        mReturnUrl = builder.mReturnUrl;

        mPaymentMethodId = builder.mPaymentMethodId;
        mPaymentMethodCreateParams = builder.mPaymentMethodCreateParams;
        mSourceId = builder.mSourceId;
        mSourceParams = builder.mSourceParams;

        mSavePaymentMethod = builder.mSavePaymentMethod;

        mExtraParams = builder.mExtraParams;
        mUseStripeSdk = builder.mShouldUseSdk;
    }

    /**
     * Create a {@link ConfirmPaymentIntentParams} without a payment method.
     */
    @NonNull
    public static ConfirmPaymentIntentParams create(
            @NonNull String clientSecret,
            @NonNull String returnUrl,
            @Nullable Map<String, Object> extraParams) {
        return new Builder(clientSecret)
                .setReturnUrl(returnUrl)
                .setExtraParams(extraParams)
                .build();
    }

    /**
     * See {@link #create(String, String, Map)}
     */
    @NonNull
    public static ConfirmPaymentIntentParams create(
            @NonNull String clientSecret,
            @NonNull String returnUrl) {
        return create(clientSecret, returnUrl, null);
    }

    /**
     * Create the parameters necessary for confirming a PaymentIntent while attaching a
     * PaymentMethod that already exits.
     *
     * @param paymentMethodId the ID of the PaymentMethod that is being attached to the
     *                        PaymentIntent being confirmed
     * @param clientSecret client secret from the PaymentIntent being confirmed
     * @param returnUrl the URL the customer should be redirected to after the authorization
     *                  process
     * @param savePaymentMethod Set to {@code true} to save this PaymentIntent’s payment method to
     *                          the associated Customer, if the payment method is not already
     *                          attached. This parameter only applies to the payment method passed
     *                          in the same request or the current payment method attached to the
     *                          PaymentIntent and must be specified again if a new payment method is
     *                          added.
     */
    @NonNull
    public static ConfirmPaymentIntentParams createWithPaymentMethodId(
            @NonNull String paymentMethodId,
            @NonNull String clientSecret,
            @Nullable String returnUrl,
            boolean savePaymentMethod,
            @Nullable Map<String, Object> extraParams) {
        return new Builder(clientSecret)
                .setPaymentMethodId(paymentMethodId)
                .setReturnUrl(returnUrl)
                .setSavePaymentMethod(savePaymentMethod)
                .setExtraParams(extraParams)
                .build();
    }

    /**
     * See {@link #createWithPaymentMethodId(String, String, String, boolean, Map)}
     */
    @NonNull
    public static ConfirmPaymentIntentParams createWithPaymentMethodId(
            @NonNull String paymentMethodId,
            @NonNull String clientSecret,
            @Nullable String returnUrl,
            boolean savePaymentMethod) {
        return createWithPaymentMethodId(paymentMethodId, clientSecret,
                returnUrl, savePaymentMethod, null);
    }

    /**
     * See {@link #createWithPaymentMethodId(String, String, String, boolean)}
     */
    @NonNull
    public static ConfirmPaymentIntentParams createWithPaymentMethodId(
            @NonNull String paymentMethodId,
            @NonNull String clientSecret,
            @Nullable String returnUrl) {
        return createWithPaymentMethodId(paymentMethodId, clientSecret,
                returnUrl, false);
    }

    /**
     * See {@link #createWithPaymentMethodId(String, String, String)}
     */
    @NonNull
    public static ConfirmPaymentIntentParams createWithPaymentMethodId(
            @NonNull String paymentMethodId,
            @NonNull String clientSecret) {
        return createWithPaymentMethodId(paymentMethodId, clientSecret, null);
    }

    /**
     * Create the parameters necessary for confirming a PaymentIntent while attaching
     * {@link PaymentMethodCreateParams} data
     *
     * @param paymentMethodCreateParams params for the PaymentMethod that will be attached to this
     *                                  PaymentIntent
     * @param clientSecret client secret from the PaymentIntent that is to be confirmed
     * @param returnUrl the URL the customer should be redirected to after the authorization
     *                  process
     * @param savePaymentMethod Set to {@code true} to save this PaymentIntent’s payment method to
     *                          the associated Customer, if the payment method is not already
     *                          attached. This parameter only applies to the payment method passed
     *                          in the same request or the current payment method attached to the
     *                          PaymentIntent and must be specified again if a new payment method is
     *                          added.
     */
    @NonNull
    public static ConfirmPaymentIntentParams createWithPaymentMethodCreateParams(
            @NonNull PaymentMethodCreateParams paymentMethodCreateParams,
            @NonNull String clientSecret,
            @Nullable String returnUrl,
            boolean savePaymentMethod,
            @Nullable Map<String, Object> extraParams) {
        return new Builder(clientSecret)
                .setPaymentMethodCreateParams(paymentMethodCreateParams)
                .setReturnUrl(returnUrl)
                .setSavePaymentMethod(savePaymentMethod)
                .setExtraParams(extraParams)
                .build();
    }

    /**
     * See {@link #createWithPaymentMethodCreateParams(PaymentMethodCreateParams, String, String,
     * boolean, Map)}
     */
    @NonNull
    public static ConfirmPaymentIntentParams createWithPaymentMethodCreateParams(
            @NonNull PaymentMethodCreateParams paymentMethodCreateParams,
            @NonNull String clientSecret,
            @Nullable String returnUrl,
            boolean savePaymentMethod) {
        return createWithPaymentMethodCreateParams(paymentMethodCreateParams,
                clientSecret, returnUrl, savePaymentMethod, null);
    }

    /**
     * See {@link #createWithPaymentMethodCreateParams(PaymentMethodCreateParams, String, String,
     * boolean)}
     */
    @NonNull
    public static ConfirmPaymentIntentParams createWithPaymentMethodCreateParams(
            @NonNull PaymentMethodCreateParams paymentMethodCreateParams,
            @NonNull String clientSecret,
            @Nullable String returnUrl) {
        return createWithPaymentMethodCreateParams(paymentMethodCreateParams,
                clientSecret, returnUrl, false);
    }

    /**
     * See {@link #createWithPaymentMethodCreateParams(PaymentMethodCreateParams, String, String)}
     */
    @NonNull
    public static ConfirmPaymentIntentParams createWithPaymentMethodCreateParams(
            @NonNull PaymentMethodCreateParams paymentMethodCreateParams,
            @NonNull String clientSecret) {
        return createWithPaymentMethodCreateParams(paymentMethodCreateParams, clientSecret, null);
    }

    /**
     * Create the parameters necessary for confirming a PaymentIntent with an
     * existing {@link Source}.
     *
     * @param sourceId the ID of the source that is being attached to the PaymentIntent being
     *                 confirmed
     * @param clientSecret client secret from the PaymentIntent being confirmed
     * @param returnUrl the URL the customer should be redirected to after the authorization
     *                  process
     * @param savePaymentMethod Set to {@code true} to save this PaymentIntent’s source to the
     *                          associated Customer, if the source is not already attached.
     *                          This parameter only applies to the source passed in the same request
     *                          or the current source attached to the PaymentIntent and must be
     *                          specified again if a new source is added.
     */
    @NonNull
    public static ConfirmPaymentIntentParams createWithSourceId(
            @NonNull String sourceId,
            @NonNull String clientSecret,
            @NonNull String returnUrl,
            boolean savePaymentMethod,
            @Nullable Map<String, Object> extraParams) {
        return new Builder(clientSecret)
                .setSourceId(sourceId)
                .setReturnUrl(returnUrl)
                .setSavePaymentMethod(savePaymentMethod)
                .setExtraParams(extraParams)
                .build();
    }

    /**
     * See {@link #createWithSourceId(String, String, String, boolean, Map)}
     */
    @NonNull
    public static ConfirmPaymentIntentParams createWithSourceId(
            @NonNull String sourceId,
            @NonNull String clientSecret,
            @NonNull String returnUrl,
            boolean savePaymentMethod) {
        return createWithSourceId(sourceId, clientSecret, returnUrl, savePaymentMethod, null);
    }

    /**
     * See {@link #createWithSourceId(String, String, String, boolean)}
     */
    @NonNull
    public static ConfirmPaymentIntentParams createWithSourceId(
            @NonNull String sourceId,
            @NonNull String clientSecret,
            @NonNull String returnUrl) {
        return createWithSourceId(sourceId, clientSecret, returnUrl, false);
    }

    /**
     * Create the parameters necessary for confirming a PaymentIntent with {@link SourceParams}
     *
     * @param sourceParams params for the source that will be attached to this PaymentIntent
     * @param clientSecret client secret from the PaymentIntent that is to be confirmed
     * @param returnUrl the URL the customer should be redirected to after the authorization
     *                  process
     * @param savePaymentMethod Set to {@code true} to save this PaymentIntent’s source to the
     *                          associated Customer, if the source is not already attached.
     *                          This parameter only applies to the source passed in the same request
     *                          or the current source attached to the PaymentIntent and must be
     *                          specified again if a new source is added.
     */
    @NonNull
    public static ConfirmPaymentIntentParams createWithSourceParams(
            @NonNull SourceParams sourceParams,
            @NonNull String clientSecret,
            @NonNull String returnUrl,
            boolean savePaymentMethod,
            @Nullable Map<String, Object> extraParams) {
        return new Builder(clientSecret)
                .setSourceParams(sourceParams)
                .setReturnUrl(returnUrl)
                .setSavePaymentMethod(savePaymentMethod)
                .setExtraParams(extraParams)
                .build();
    }

    @NonNull
    public static ConfirmPaymentIntentParams createWithSourceParams(
            @NonNull SourceParams sourceParams,
            @NonNull String clientSecret,
            @NonNull String returnUrl,
            boolean savePaymentMethod) {
        return createWithSourceParams(sourceParams, clientSecret, returnUrl, savePaymentMethod,
                null);
    }

    @NonNull
    public static ConfirmPaymentIntentParams createWithSourceParams(
            @NonNull SourceParams sourceParams,
            @NonNull String clientSecret,
            @NonNull String returnUrl) {
        return createWithSourceParams(sourceParams, clientSecret, returnUrl, false);
    }

    @Nullable
    public PaymentMethodCreateParams getPaymentMethodCreateParams() {
        return mPaymentMethodCreateParams;
    }

    @Nullable
    public String getPaymentMethodId() {
        return mPaymentMethodId;
    }

    @Nullable
    public SourceParams getSourceParams() {
        return mSourceParams;
    }

    @Nullable
    public String getSourceId() {
        return mSourceId;
    }

    @Nullable
    public Map<String, Object> getExtraParams() {
        return mExtraParams;
    }

    @NonNull
    public String getClientSecret() {
        return mClientSecret;
    }

    @Nullable
    public String getReturnUrl() {
        return mReturnUrl;
    }

    public boolean shouldSavePaymentMethod() {
        return mSavePaymentMethod;
    }

    @Override
    public boolean shouldUseStripeSdk() {
        return mUseStripeSdk;
    }

    @NonNull
    @Override
    public ConfirmPaymentIntentParams withShouldUseStripeSdk(boolean shouldUseStripeSdk) {
        return toBuilder()
                .setShouldUseSdk(shouldUseStripeSdk)
                .build();
    }

    /**
     * Create a Map representing this object that is prepared for the Stripe API.
     */
    @NonNull
    public Map<String, Object> toParamMap() {
        final AbstractMap<String, Object> networkReadyMap = new HashMap<>();

        if (mPaymentMethodCreateParams != null) {
            networkReadyMap.put(API_PARAM_PAYMENT_METHOD_DATA,
                    mPaymentMethodCreateParams.toParamMap());
        } else if (mPaymentMethodId != null) {
            networkReadyMap.put(API_PARAM_PAYMENT_METHOD_ID, mPaymentMethodId);
        } else if (mSourceParams != null) {
            networkReadyMap.put(API_PARAM_SOURCE_DATA, mSourceParams.toParamMap());
        } else if (mSourceId != null) {
            networkReadyMap.put(API_PARAM_SOURCE_ID, mSourceId);
        }

        if (mReturnUrl != null) {
            networkReadyMap.put(API_PARAM_RETURN_URL, mReturnUrl);
        }
        networkReadyMap.put(API_PARAM_CLIENT_SECRET, mClientSecret);
        if (mExtraParams != null) {
            networkReadyMap.putAll(mExtraParams);
        }

        if (mSavePaymentMethod) {
            networkReadyMap.put(API_PARAM_SAVE_PAYMENT_METHOD, true);
        }

        if (mUseStripeSdk) {
            networkReadyMap.put(API_PARAM_USE_STRIPE_SDK, true);
        }

        return networkReadyMap;
    }

    @NonNull
    private Builder toBuilder() {
        return new Builder(mClientSecret)
                .setReturnUrl(mReturnUrl)
                .setPaymentMethodId(mPaymentMethodId)
                .setPaymentMethodCreateParams(mPaymentMethodCreateParams)
                .setSourceId(mSourceId)
                .setSourceParams(mSourceParams)
                .setSavePaymentMethod(mSavePaymentMethod)
                .setExtraParams(mExtraParams);
    }

    private static final class Builder implements ObjectBuilder<ConfirmPaymentIntentParams> {
        @NonNull private final String mClientSecret;

        @Nullable private PaymentMethodCreateParams mPaymentMethodCreateParams;
        @Nullable private String mPaymentMethodId;
        @Nullable private SourceParams mSourceParams;
        @Nullable private String mSourceId;

        @Nullable private Map<String, Object> mExtraParams;
        @Nullable private String mReturnUrl;

        private boolean mSavePaymentMethod;
        private boolean mShouldUseSdk;

        /**
         * Sets the client secret that is used to authenticate actions on this PaymentIntent
         * @param clientSecret client secret associated with this PaymentIntent
         */
        private Builder(@NonNull String clientSecret) {
            mClientSecret = Objects.requireNonNull(clientSecret);
        }

        /**
         * Sets the PaymentMethod data that will be included with this PaymentIntent
         *
         * @param paymentMethodCreateParams Params for the PaymentMethod that will be attached to
         *                                  this PaymentIntent. Only one of PaymentMethodParam,
         *                                  PaymentMethodId, SourceParam, SourceId should be used
         *                                  at a time.
         */
        @NonNull
        private Builder setPaymentMethodCreateParams(
                @NonNull PaymentMethodCreateParams paymentMethodCreateParams) {
            this.mPaymentMethodCreateParams = paymentMethodCreateParams;
            return this;
        }

        /**
         * Sets a pre-existing PaymentMethod that will be attached to this PaymentIntent
         *
         * @param paymentMethodId The ID of the PaymentMethod that is being attached to this
         *                        PaymentIntent. Only one of PaymentMethodParam, PaymentMethodId,
         *                        SourceParam, SourceId should be used at a time.
         */
        @NonNull
        private Builder setPaymentMethodId(@NonNull String paymentMethodId) {
            mPaymentMethodId = paymentMethodId;
            return this;
        }

        /**
         * Sets the source data that will be included with this PaymentIntent
         *
         * @param sourceParams params for the source that will be attached to this PaymentIntent.
         *                     Only one of SourceParam and SourceId should be used at at time.
         */
        @NonNull
        private Builder setSourceParams(@NonNull SourceParams sourceParams) {
            mSourceParams = sourceParams;
            return this;
        }

        /**
         * Sets a pre-existing source that will be attached to this PaymentIntent
         * @param sourceId the ID of an existing Source that is being attached to this
         *                 PaymentIntent. Only one of SourceParam and SourceId should be used at a
         *                 time.
         */
        @NonNull
        private Builder setSourceId(@Nullable String sourceId) {
            mSourceId = sourceId;
            return this;
        }

        /**
         * @param returnUrl the URL the customer should be redirected to after the authentication
         *                  process
         */
        @NonNull
        private Builder setReturnUrl(@Nullable String returnUrl) {
            mReturnUrl = returnUrl;
            return this;
        }

        /**
         * @param extraParams params that will be included in the request. Incorrect params may
         *                    result in errors when connecting to Stripe's API.
         */
        @NonNull
        private Builder setExtraParams(@Nullable Map<String, Object> extraParams) {
            mExtraParams = extraParams;
            return this;
        }

        @NonNull
        private Builder setSavePaymentMethod(boolean savePaymentMethod) {
            mSavePaymentMethod = savePaymentMethod;
            return this;
        }

        @NonNull
        private Builder setShouldUseSdk(boolean shouldUseSdk) {
            mShouldUseSdk = shouldUseSdk;
            return this;
        }

        @NonNull
        @Override
        public ConfirmPaymentIntentParams build() {
            return new ConfirmPaymentIntentParams(this);
        }
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        return this == obj || (obj instanceof ConfirmPaymentIntentParams &&
                typedEquals((ConfirmPaymentIntentParams) obj));
    }

    private boolean typedEquals(@NonNull ConfirmPaymentIntentParams params) {
        return ObjectUtils.equals(mReturnUrl, params.mReturnUrl)
                && ObjectUtils.equals(mClientSecret, params.mClientSecret)
                && ObjectUtils.equals(mPaymentMethodId, params.mPaymentMethodId)
                && ObjectUtils.equals(mPaymentMethodCreateParams, params.mPaymentMethodCreateParams)
                && ObjectUtils.equals(mSourceId, params.mSourceId)
                && ObjectUtils.equals(mSourceParams, params.mSourceParams)
                && ObjectUtils.equals(mExtraParams, params.mExtraParams)
                && ObjectUtils.equals(mSavePaymentMethod, params.mSavePaymentMethod)
                && ObjectUtils.equals(mUseStripeSdk, params.mUseStripeSdk);
    }

    @Override
    public int hashCode() {
        return ObjectUtils.hash(mReturnUrl, mClientSecret, mPaymentMethodId,
                mPaymentMethodCreateParams, mSourceId, mSourceParams, mExtraParams,
                mSavePaymentMethod, mUseStripeSdk);
    }
}

File: stripe/src/main/java/com/stripe/android/model/ConfirmPaymentIntentParams.kt
package com.stripe.android.model

import android.os.Parcelable
import com.stripe.android.model.ConfirmPaymentIntentParams.SetupFutureUsage
import com.stripe.android.model.ConfirmPaymentIntentParams.Shipping
import com.stripe.android.model.ConfirmStripeIntentParams.Companion.PARAM_CLIENT_SECRET
import com.stripe.android.model.ConfirmStripeIntentParams.Companion.PARAM_MANDATE_ID
import com.stripe.android.model.ConfirmStripeIntentParams.Companion.PARAM_PAYMENT_METHOD_DATA
import com.stripe.android.model.ConfirmStripeIntentParams.Companion.PARAM_PAYMENT_METHOD_ID
import com.stripe.android.model.ConfirmStripeIntentParams.Companion.PARAM_RETURN_URL
import com.stripe.android.model.ConfirmStripeIntentParams.Companion.PARAM_USE_STRIPE_SDK
import kotlinx.android.parcel.Parcelize

data class ConfirmPaymentIntentParams internal constructor(
    val paymentMethodCreateParams: PaymentMethodCreateParams? = null,

    /**
     * ID of the payment method (a PaymentMethod, Card, or compatible Source object) to attach to
     * this PaymentIntent.
     *
     * [payment_method](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-payment_method)
     */
    val paymentMethodId: String? = null,
    val sourceParams: SourceParams? = null,
    val sourceId: String? = null,

    val extraParams: Map<String, Any>? = null,

    /**
     * The client secret of this PaymentIntent. Used for client-side retrieval using a
     * publishable key.
     *
     * The client secret can be used to complete a payment from your frontend. It should not be
     * stored, logged, embedded in URLs, or exposed to anyone other than the customer. Make sure
     * that you have TLS enabled on any page that includes the client secret.
     *
     * Refer to our docs to [accept a payment](https://stripe.com/docs/payments/accept-a-payment)
     * and learn about how `client_secret` should be handled.
     *
     * [client_secret](https://stripe.com/docs/api/payment_intents/object#payment_intent_object-client_secret)
     */
    override val clientSecret: String,

    /**
     * The URL to redirect your customer back to after they authenticate or cancel their payment on
     * the payment method’s app or site. If you’d prefer to redirect to a mobile application, you
     * can alternatively supply an application URI scheme. This parameter is only used for cards
     * and other redirect-based payment methods.
     *
     * [return_url](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-return_url)
     */
    val returnUrl: String? = null,

    /**
     * If the PaymentIntent has a `payment_method` and a `customer` or if you’re attaching a payment
     * method to the PaymentIntent in this request, you can pass `save_payment_method=true` to save
     * the payment method to the customer. Defaults to `false`.
     *
     * If the payment method is already saved to a customer, this does nothing. If this type of
     * payment method cannot be saved to a customer, the request will error.
     *
     * [save_payment_method](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-save_payment_method)
     */
    private val savePaymentMethod: Boolean? = null,

    /**
     * Set to `true` only when using manual confirmation and the iOS or Android SDKs to handle
     * additional authentication steps.
     *
     * [use_stripe_sdk](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-use_stripe_sdk)
     */
    private val useStripeSdk: Boolean = false,

    /**
     * Payment-method-specific configuration for this PaymentIntent.
     *
     * [payment_method_options](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-payment_method_options)
     */
    private val paymentMethodOptions: PaymentMethodOptionsParams? = null,

    /**
     * ID of the mandate to be used for this payment.
     *
     * [mandate](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-mandate)
     */
    private val mandateId: String? = null,

    /**
     * This hash contains details about the Mandate to create.
     *
     * [mandate_data](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-mandate_data)
     */
    private val mandateData: MandateDataParams? = null,

    /**
     * See [SetupFutureUsage]
     *
     * [setup_future_usage](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-setup_future_usage)
     */
    private val setupFutureUsage: SetupFutureUsage? = null,

    /**
     * See [Shipping]
     *
     * [shipping](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-shipping)
     */
    private val shipping: Shipping? = null
) : ConfirmStripeIntentParams {

    fun shouldSavePaymentMethod(): Boolean {
        return savePaymentMethod == true
    }

    override fun shouldUseStripeSdk(): Boolean {
        return useStripeSdk
    }

    override fun withShouldUseStripeSdk(shouldUseStripeSdk: Boolean): ConfirmPaymentIntentParams {
        return copy(useStripeSdk = shouldUseStripeSdk)
    }

    /**
     * Create a Map representing this object that is prepared for the Stripe API.
     */
    override fun toParamMap(): Map<String, Any> {
        return mapOf(
            PARAM_CLIENT_SECRET to clientSecret,
            PARAM_USE_STRIPE_SDK to useStripeSdk
        ).plus(
            savePaymentMethod?.let {
                mapOf(PARAM_SAVE_PAYMENT_METHOD to it)
            }.orEmpty()
        ).plus(
            mandateId?.let { mapOf(PARAM_MANDATE_ID to it) }.orEmpty()
        ).plus(
            mandateDataParams?.let {
                mapOf(ConfirmStripeIntentParams.PARAM_MANDATE_DATA to it)
            }.orEmpty()
        ).plus(
            returnUrl?.let { mapOf(PARAM_RETURN_URL to it) }.orEmpty()
        ).plus(
            paymentMethodOptions?.let {
                mapOf(PARAM_PAYMENT_METHOD_OPTIONS to it.toParamMap())
            }.orEmpty()
        ).plus(
            setupFutureUsage?.let {
                mapOf(PARAM_SETUP_FUTURE_USAGE to it.code)
            }.orEmpty()
        ).plus(
            shipping?.let {
                mapOf(PARAM_SHIPPING to shipping.toParamMap())
            }.orEmpty()
        ).plus(
            paymentMethodParamMap
        ).plus(
            extraParams.orEmpty()
        )
    }

    private val paymentMethodParamMap: Map<String, Any>
        get() {
            return when {
                paymentMethodCreateParams != null -> {
                    mapOf(PARAM_PAYMENT_METHOD_DATA to paymentMethodCreateParams.toParamMap())
                }
                paymentMethodId != null -> {
                    mapOf(PARAM_PAYMENT_METHOD_ID to paymentMethodId)
                }
                sourceParams != null -> {
                    mapOf(PARAM_SOURCE_DATA to sourceParams.toParamMap())
                }
                sourceId != null -> {
                    mapOf(PARAM_SOURCE_ID to sourceId)
                }
                else -> emptyMap()
            }
        }

    /**
     * Use the user-defined [MandateDataParams] if specified, otherwise create a default
     * [MandateDataParams] if necessary.
     */
    private val mandateDataParams: Map<String, Any>?
        get() {
            return mandateData?.toParamMap()
                ?: if (paymentMethodCreateParams?.type?.hasMandate == true && mandateId == null) {
                    // Populate with default "online" MandateData
                    MandateDataParams(MandateDataParams.Type.Online.DEFAULT).toParamMap()
                } else {
                    null
                }
        }

    /**
     * [setup_future_usage](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-setup_future_usage)
     *
     * Use `on_session` if you intend to only reuse the payment method when your customer is
     * present in your checkout flow. Use `off_session` if your customer may or may not be in
     * your checkout flow.
     *
     * Stripe uses `setup_future_usage` to dynamically optimize your payment flow and comply with
     * regional legislation and network rules. For example, if your customer is impacted by
     * [SCA](https://stripe.com/docs/strong-customer-authentication), using `off_session` will
     * ensure that they are authenticated while processing this PaymentIntent. You will then be
     * able to collect [off-session payments](https://stripe.com/docs/payments/cards/charging-saved-cards#off-session-payments-with-saved-cards)
     * for this customer.
     *
     * If `setup_future_usage` is already set, you may only update the value
     * from `on_session` to `off_session`.
     */
    enum class SetupFutureUsage(
        internal val code: String
    ) {
        /**
         * Use `on_session` if you intend to only reuse the payment method when your customer
         * is present in your checkout flow.
         */
        OnSession("on_session"),

        /**
         * Use `off_session` if your customer may or may not be in your checkout flow.
         */
        OffSession("off_session")
    }

    /**
     * Shipping information for this PaymentIntent.
     *
     * [shipping](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-shipping)
     */
    @Parcelize
    data class Shipping @JvmOverloads constructor(
        /**
         * [shipping.address](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-shipping-address)
         *
         * Shipping address.
         */
        internal val address: Address,

        /**
         * [shipping.name](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-shipping-name)
         *
         * Recipient name.
         */
        internal val name: String,

        /**
         * [shipping.carrier](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-shipping-carrier)
         *
         * The delivery service that shipped a physical product, such as Fedex, UPS, USPS, etc.
         */
        internal val carrier: String? = null,

        /**
         * [shipping.phone](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-shipping-phone)
         *
         * Recipient phone (including extension).
         */
        internal val phone: String? = null,

        /**
         * [shipping.tracking_number](https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-shipping-tracking_number)
         *
         * The tracking number for a physical product, obtained from the delivery service.
         * If multiple tracking numbers were generated for this purchase, please separate
         * them with commas.
         */
        internal val trackingNumber: String? = null
    ) : StripeParamsModel, Parcelable {
        override fun toParamMap(): Map<String, Any> {
            return listOf(
                PARAM_ADDRESS to address.toParamMap(),
                PARAM_NAME to name,
                PARAM_CARRIER to carrier,
                PARAM_PHONE to phone,
                PARAM_TRACKING_NUMBER to trackingNumber
            ).fold(emptyMap()) { acc, (key, value) ->
                acc.plus(
                    value?.let { mapOf(key to it) }.orEmpty()
                )
            }
        }

        private companion object {
            private const val PARAM_ADDRESS = "address"
            private const val PARAM_NAME = "name"
            private const val PARAM_CARRIER = "carrier"
            private const val PARAM_PHONE = "phone"
            private const val PARAM_TRACKING_NUMBER = "tracking_number"
        }
    }

    companion object {
        const val PARAM_SOURCE_DATA: String = "source_data"

        internal const val PARAM_SOURCE_ID = "source"
        internal const val PARAM_SAVE_PAYMENT_METHOD = "save_payment_method"
        internal const val PARAM_PAYMENT_METHOD_OPTIONS = "payment_method_options"
        private const val PARAM_SETUP_FUTURE_USAGE = "setup_future_usage"
        private const val PARAM_SHIPPING = "shipping"

        /**
         * Create a [ConfirmPaymentIntentParams] without a payment method.
         */
        @JvmOverloads
        @JvmStatic
        fun create(
            clientSecret: String,
            returnUrl: String? = null,
            extraParams: Map<String, Any>? = null,
            shipping: Shipping? = null
        ): ConfirmPaymentIntentParams {
            return ConfirmPaymentIntentParams(
                clientSecret = clientSecret,
                returnUrl = returnUrl,
                extraParams = extraParams,
                shipping = shipping
            )
        }

        /**
         * Create the parameters necessary for confirming a PaymentIntent while attaching a
         * PaymentMethod that already exits.
         *
         * @param paymentMethodId the ID of the PaymentMethod that is being attached to the
         * PaymentIntent being confirmed
         * @param clientSecret client secret from the PaymentIntent being confirmed
         * @param returnUrl the URL the customer should be redirected to after the authorization
         * process
         * @param savePaymentMethod Set to `true` to save this PaymentIntent’s payment method to
         * the associated Customer, if the payment method is not already
         * attached. This parameter only applies to the payment method passed
         * in the same request or the current payment method attached to the
         * PaymentIntent and must be specified again if a new payment method is
         * added.
         * @param paymentMethodOptions Optional [PaymentMethodOptionsParams]
         * @param mandateId Optional ID of the Mandate to be used for this payment.
         * @param mandateData Optional details about the Mandate to create.
         * @param setupFutureUsage Optional. See [SetupFutureUsage].
         * @param shipping Optional. See [Shipping].
         */
        @JvmOverloads
        @JvmStatic
        fun createWithPaymentMethodId(
            paymentMethodId: String,
            clientSecret: String,
            returnUrl: String? = null,
            savePaymentMethod: Boolean? = null,
            extraParams: Map<String, Any>? = null,
            paymentMethodOptions: PaymentMethodOptionsParams? = null,
            mandateId: String? = null,
            mandateData: MandateDataParams? = null,
            setupFutureUsage: SetupFutureUsage? = null,
            shipping: Shipping? = null
        ): ConfirmPaymentIntentParams {
            return ConfirmPaymentIntentParams(
                clientSecret = clientSecret,
                paymentMethodId = paymentMethodId,
                returnUrl = returnUrl,
                savePaymentMethod = savePaymentMethod,
                extraParams = extraParams,
                paymentMethodOptions = paymentMethodOptions,
                mandateId = mandateId,
                mandateData = mandateData,
                setupFutureUsage = setupFutureUsage,
                shipping = shipping
            )
        }

        /**
         * Create the parameters necessary for confirming a PaymentIntent while attaching
         * [PaymentMethodCreateParams] data
         *
         * @param paymentMethodCreateParams params for the PaymentMethod that will be attached to this
         * PaymentIntent
         * @param clientSecret client secret from the PaymentIntent that is to be confirmed
         * @param returnUrl the URL the customer should be redirected to after the authorization
         * process
         * @param savePaymentMethod Set to `true` to save this PaymentIntent’s payment method to
         * the associated Customer, if the payment method is not already
         * attached. This parameter only applies to the payment method passed
         * in the same request or the current payment method attached to the
         * PaymentIntent and must be specified again if a new payment method is
         * added.
         * @param mandateId optional ID of the Mandate to be used for this payment.
         * @param mandateData optional details about the Mandate to create.
         * @param setupFutureUsage Optional. See [SetupFutureUsage].
         * @param shipping Optional. See [Shipping].
         */
        @JvmOverloads
        @JvmStatic
        fun createWithPaymentMethodCreateParams(
            paymentMethodCreateParams: PaymentMethodCreateParams,
            clientSecret: String,
            returnUrl: String? = null,
            savePaymentMethod: Boolean? = null,
            extraParams: Map<String, Any>? = null,
            mandateId: String? = null,
            mandateData: MandateDataParams? = null,
            setupFutureUsage: SetupFutureUsage? = null,
            shipping: Shipping? = null
        ): ConfirmPaymentIntentParams {
            return ConfirmPaymentIntentParams(
                clientSecret = clientSecret,
                paymentMethodCreateParams = paymentMethodCreateParams,
                returnUrl = returnUrl,
                savePaymentMethod = savePaymentMethod,
                extraParams = extraParams,
                mandateId = mandateId,
                mandateData = mandateData,
                setupFutureUsage = setupFutureUsage,
                shipping = shipping
            )
        }

        /**
         * Create the parameters necessary for confirming a PaymentIntent with an
         * existing [Source].
         *
         * @param sourceId the ID of the source that is being attached to the PaymentIntent being
         * confirmed
         * @param clientSecret client secret from the PaymentIntent being confirmed
         * @param returnUrl the URL the customer should be redirected to after the authorization
         * process
         * @param savePaymentMethod Set to `true` to save this PaymentIntent’s source to the
         * associated Customer, if the source is not already attached.
         * This parameter only applies to the source passed in the same request
         * or the current source attached to the PaymentIntent and must be
         * specified again if a new source is added.
         * @param shipping Optional. See [Shipping].
         */
        @JvmOverloads
        @JvmStatic
        fun createWithSourceId(
            sourceId: String,
            clientSecret: String,
            returnUrl: String,
            savePaymentMethod: Boolean? = null,
            extraParams: Map<String, Any>? = null,
            shipping: Shipping? = null
        ): ConfirmPaymentIntentParams {
            return ConfirmPaymentIntentParams(
                clientSecret = clientSecret,
                sourceId = sourceId,
                returnUrl = returnUrl,
                savePaymentMethod = savePaymentMethod,
                extraParams = extraParams,
                shipping = shipping
            )
        }

        /**
         * Create the parameters necessary for confirming a PaymentIntent with [SourceParams]
         *
         * @param sourceParams params for the source that will be attached to this PaymentIntent
         * @param clientSecret client secret from the PaymentIntent that is to be confirmed
         * @param returnUrl the URL the customer should be redirected to after the authorization
         * process
         * @param savePaymentMethod Set to `true` to save this PaymentIntent’s source to the
         * associated Customer, if the source is not already attached.
         * This parameter only applies to the source passed in the same request
         * or the current source attached to the PaymentIntent and must be
         * specified again if a new source is added.
         * @param shipping Optional. See [Shipping].
         */
        @JvmOverloads
        @JvmStatic
        fun createWithSourceParams(
            sourceParams: SourceParams,
            clientSecret: String,
            returnUrl: String,
            savePaymentMethod: Boolean? = null,
            extraParams: Map<String, Any>? = null,
            shipping: Shipping? = null
        ): ConfirmPaymentIntentParams {
            return ConfirmPaymentIntentParams(
                clientSecret = clientSecret,
                sourceParams = sourceParams,
                returnUrl = returnUrl,
                savePaymentMethod = savePaymentMethod,
                extraParams = extraParams,
                shipping = shipping
            )
        }
    }
}
